### ISSUE TITLE
Functionality system

### ISSUE THREAD

**Issue by vasilepi** [vasilepi/Thesis#1]
Design a high-level functionality system. Should include:
- Sensors
- Control algorithms

## To do's
- [x] Declare sensor types and number of sensors
- [x] Design control systems for basic robot functions

The robot should be able to achieve the below with its functionalities.
## Objective
- Locate victims
- Obstacle recognition/negotiation
- Localization and mapping of environment
- Autonomous speed control

**Comment by vasilepi** [comment]
The sensor that are proposed are:
- x1 Thermal Camera with wide FOV (>80deg)
- x2 Bidirectional Microphones
- x1 Vibration sensor
- x1 2D 360 FOV LiDAR
- x1 IMU
- x1 PSD
- x4 CO2 sensors

Positions:
- The camera will be on a base at the top surface of the robot body so it can have a clear field of view
- The microphone will be near the camera on the base facing perpendicular directions directions so the bipolarity of the two mics will cover a 360deg field.
<img src="https://github.com/user-attachments/assets/ffdbd1ca-56a3-4b57-ac63-2d6451f60760" width="300" height="300">

- The vibration sensor should be at the bottom surface of the robot body, as close to the ground as possible so the input will be as intact as it can be due to damping

- The LiDAR should be above the robot's highest surface so it can scan only the environment and not the robot parts
 
- The IMU should as close to the robot's COG as possible
- The PSD should be as close to the ground as possible ans as close to the front wheels as possible
- The CO2 will be on each vertical side of the robot body

### SUMMING UP:
![image](https://github.com/user-attachments/assets/5cbbf4f9-bca5-4394-a35c-6e89adbfb9d8)

**Comment by vasilepi** [comment]
The robot is assumed to work semi-autonomously. That means that whenever needed a human operator can take control of the robot. For the functionality it is assumed that the robot works autonomously until the operator intervene. The functionality of the above sensor picks is justified by the following proposed automation algorithm for victim identification. The following algorithm proves that the sensors chosen can function adequately for a USAR mission. Here for every sensor it is assumed that after collecting the data, an appropriate NN algorithm is used to determine whether or not a victim is identified (by that sensor reading only). So:
- A: the response of the  Camera Sensor NN algorithm. 
- Mi: the response of the Microphones NN algorithm. Before the NN algorithm the readings are filtered properly.
- Ci: the response of the CO2 sensors NN algorithm
- V: the response of the Vibe sensor NN algorithm. Before the NN algorithm the readings are filtered properly.

For each of the above, the result will be a boolean variable where 1 means that the algorithm identified a victim and 0 that it didn't. For the Ci and Mi sensor results, at least one of them needs to provide a result. That's why these sensor pass through a LOGICAL gate (Each sensor reading passes through the NN algorithm. So for 2 microphones we have 2 readings and 2 passes through the NN algorithm so 2 results). Also:
- S1: Scenario 1 is when the system prompts the human operator to call the decision (whether or not to keep searching, if the human operator detects a victim through the camera, to check if the victim identified is actually a victim and various other functions)
- VI: stands for Victim identification through the automation algorithm. Every time the algorithm itself identifies a victim without the need of the operator, the operator is notified to confirm the decision
- S2: Scenario 2 is the vibration localization algorithm. It stands for a localization algorithm where the robot tracks the vibrations source from the sensor readings
- SL: Sound localization algorithm
- 0: resume search, ignore sensor readings

### RECAP
| Variable | Info |
|---| --|
| A| A=1 or A=0 means the camera sensor NN algorithm identified or did not identified a victim|
| Mi| Mi=1 or Mi=0 means the microphone sensor i NN algorithm identified or did not identified a victim|
| Ci| Ci=1 or Ci=0 means the CO2 sensor i NN algorithm identified or did not identified a victim|
| V| V=1 or V=0 means the vibration sensor NN algorithm identified or did not identified a victim|
| S1| Prompt to the human operator to decide the next robot funciton |
| S2| Initiate vibrations localization algorithm|
| SL| Initiate sound localization algorithm|
|0| Resume search|
| VI|Victim identified by the automation algorithm alone|
| S1, S2| Notify the operator to start S2 |
| S1, 0| Notify the operator to resume search |


![image](https://github.com/user-attachments/assets/a78add5e-bbb4-40cf-a907-0ed7a33c250f)

**Comment by vasilepi** [comment]
# Changes

## In order for the robot to work semi-autonomously
- ROS will be used for a publish/subscribe system
- Master node should be installed on the RPi
- The RPi should also be configured as a Wi-Fi Host

In order for the operator to control the robot, a laptop with LINUX OS will be the interface for the sensor data information exchange. The operator will also use a PS4 controller to control the robot actuators if needed. The system architecture following the subscribe/publish principal will be configured as follows:
- ROS master on RPi
- Each sensor will utilize a small ROS node that publishes its measurements to topics
- The motor drivers will be subscribed to a topic that will receive commands from the PS4 gamepad (publish)

<img width="1160" height="750" alt="Image" src="https://github.com/user-attachments/assets/b95ecf59-9e94-4367-91e3-30c96b61e9b3" />



### Electronics BOM can be found [here](https://docs.google.com/spreadsheets/d/1dSLtLHoS9FY3o4XzIignpxQ59i9sujoK4btNogsYrk4/edit?gid=2089421684#gid=2089421684).

- Li-Ion battery pack of 100Ah and 48V will be used
- The CO2 sensors with [slow response](https://grobotronics.com/gravity-analog-co2-gas-sensor-mg-811.html?sl=en) will be placed on the sides of the robot while the [faster](https://www.co2meter.com/products/k-30-fr-co2-sensor) will be placed on the front and rear of robot.

## _The robot will have one minute stall condition for the fast response CO2 sensor to warm-up according to the product datasheet. After that the robot can start to feed data to the operator and once the slow response CO2 sensors warm-up as well they should start sending data too._

- The microphone capsule will be soldered to a male jack interface. The jack will be then plugged into a usb sound card which will be plugged (through usb) to the MCU
- The ODrive S1 for the BLDC motors will be powered through the RPi 5V GPIO
- The IMU, PSD and VIBE sensors will utilize the RPi GPIO pins. The rest of the sensors will be plugged in with USB

<img width="1442" height="863" alt="Image" src="https://github.com/user-attachments/assets/dcc822f4-46b6-4a78-87ea-443de58c2d28" />
